File: backend/radioai/showmeta.py
import json
from datetime import datetime, timezone


def build_segments(items, total_s):
    """Sort song markers by start_s and fill end_s (next start, last=total)."""
    ordered = sorted(items, key=lambda x: x["start_s"])
    out = []
    for idx, it in enumerate(ordered):
        end = ordered[idx + 1]["start_s"] if idx + 1 < len(ordered) else total_s
        out.append({**it, "end_s": round(float(end), 2)})
    return out


def write_show_json(path, *, station, dj, duration_s, segments, talk,
                    generated_at=None):
    meta = {
        "station": station,
        "dj": dj,
        "duration_s": round(float(duration_s), 2),
        "generated_at": generated_at or datetime.now(timezone.utc).isoformat(),
        "segments": segments,
        "talk": talk,
    }
    with open(path, "w", encoding="utf-8") as f:
        json.dump(meta, f, ensure_ascii=False, indent=2)
    return meta

File: android/station/src/main/java/ai/kolai/station/ShowMeta.kt
package ai.kolai.station

/**
 * Show-meta segment building, ported 1:1 from backend/radioai/showmeta.py
 * (build_segments). The Python build_segments takes a list of song-marker dicts
 * ({"type","title","artist","start_s"}) and returns the same dicts with an
 * "end_s" filled in (next marker's start_s; the last one gets total_s). We model
 * the input/output as typed data classes mirroring those dict keys.
 *
 * write_show_json (JSON serialization to disk) is NOT ported here; it is a
 * device/serving concern handled elsewhere.
 *
 * Rounding note: Python round(x, 2) is round-half-to-even; mirror with Math.rint
 * (x * 100) / 100, matching the rest of the engine (Dsp).
 */

/** A song marker before end_s is filled. Mirrors the Python input dict
 *  {"type":"song","title","artist","start_s"} (type is implicit/constant). */
data class SongEvent(
    val title: String,
    val artist: String,
    val startS: Double,
)

/** A finished segment. Mirrors the Python output dict
 *  {"type","title","artist","start_s","end_s"} plus a stable ordinal [index]
 *  (the serve layer / UI uses the post-sort position). */
data class Segment(
    val index: Int,
    val title: String,
    val artist: String,
    val startS: Double,
    val endS: Double,
)

/**
 * Python: build_segments. Sort song markers by start_s, then fill end_s with the
 * next marker's start_s (the last marker gets total_s), each rounded to 2 dp.
 */
fun buildSegments(items: List<SongEvent>, totalS: Double): List<Segment> {
    val ordered = items.sortedBy { it.startS }
    val out = ArrayList<Segment>(ordered.size)
    for (idx in ordered.indices) {
        val it = ordered[idx]
        val end = if (idx + 1 < ordered.size) ordered[idx + 1].startS else totalS
        out.add(
            Segment(
                index = idx,
                title = it.title,
                artist = it.artist,
                startS = it.startS,
                endS = round2(end),
            )
        )
    }
    return out
}

/** Python round(x, 2): round-half-to-even to 2 decimal places. */
internal fun round2(x: Double): Double = Math.rint(x * 100.0) / 100.0